execution-engine/src/wasi/common.rs

```diff
@@ -0,0 +1,1519 @@
+    /// interact with the execution engine.                  
+    #[inline]
+    pub(crate) fn clock_time_get(
+        &mut self,
+        memory_ref: &mut impl MemoryHandler,
+        clock_id: u32,
+        precision: u64,
+        address: u32,
+    ) -> ErrNo {
+        let clock_id = decode_wasi_arg!(clock_id, u8);
+        let clock_id = decode_wasi_arg!(clock_id, ClockId);
+        let fs = lock_vfs!(self);
+        match fs.clock_time_get(clock_id, precision.into()) {
+            Ok(o) => memory_ref.write_buffer(address, &u64::to_le_bytes(o.as_nanos())),
+            Err(e) => e,
+        }
+    }
+
+    /// The implementation of the WASI `fd_advise` function. It requires an extra `memory_ref` to
+    /// interact with the execution engine.                  
+    #[inline]
+    pub(crate) fn fd_advise(
+        &mut self,
+        _: &mut impl MemoryHandler,
+        fd: u32,
+        offset: u64,
+        len: u64,
+        advice: u8,
+    ) -> ErrNo {
+        let advice = decode_wasi_arg!(advice, Advice);
+        let mut fs = lock_vfs!(self);
+        if let Err(err) = fs.fd_advise(fd.into(), offset, len, advice) {
+            return err;
+        }
+        return ErrNo::Success;
+    }
+
+    /// The implementation of the WASI `fd_allocate` function. It requires an extra `memory_ref` to
+    /// interact with the execution engine.                  
+    #[inline]
+    pub(crate) fn fd_allocate(
+        &mut self,
+        _: &mut impl MemoryHandler,
+        fd: u32,
+        offset: u64,
+        len: u64,
+    ) -> ErrNo {
+        let mut fs = lock_vfs!(self);
+        match fs.fd_allocate(fd.into(), offset, len) {
+            Ok(_) => ErrNo::Success,
+            Err(e) => e,
+        }
+    }
+
+    /// The implementation of the WASI `fd_close` function. It requires an extra `memory_ref` to
+    /// interact with the execution engine.                  
+    #[inline]
+    pub(crate) fn fd_close(&mut self, fd: u32) -> ErrNo {
+        let mut fs = lock_vfs!(self);
+        match fs.fd_close(fd.into()) {
+            Ok(_) => ErrNo::Success,
+            Err(e) => e,
+        }
+    }
+
+    /// The implementation of the WASI `fd_datasync` function. It requires an extra `memory_ref` to
+    /// interact with the execution engine.                  
+    #[inline]
+    pub(crate) fn fd_datasync(&mut self, _: &mut impl MemoryHandler, fd: u32) -> ErrNo {
+        let mut fs = lock_vfs!(self);
+        match fs.fd_datasync(fd.into()) {
+            Ok(_) => ErrNo::Success,
+            Err(e) => e,
+        }
+    }
+
+    /// The implementation of the WASI `fd_fdstat_get` function. It requires an extra `memory_ref` to
+    /// interact with the execution engine.                  
+    #[inline]
+    pub(crate) fn fd_fdstat_get(
+        &self,
+        memory_ref: &mut impl MemoryHandler,
+        fd: u32,
+        address: u32,
+    ) -> ErrNo {
+        let fs = lock_vfs!(self);
+        match fs.fd_fdstat_get(fd.into()) {
```

Comment:
Maybe neater in terms of `map` and `map_err`?

Reply:
Have reworked here by using `?`.